### PR TITLE
add building workflow for TE/Jax

### PR DESCRIPTION
**Pull request by jeng1220**
This is the first commit for TE/Jax.

The build process for TE/PyTorch and TE/Jax components is different. In the building step, the TE/Jax only depends on CPP STL and CUDA, so we simply add needed building commands in CMakeList.txt. Unlike PyTorch, Jax has nothing like `torch.utils.cpp_extension.BuildExtension` and `torch.utils.cpp_extension.CUDAExtension`.

To ensure that users can build only TE/PyTorch, only TE/Jax, or both, I created `FrameworkBuilder` to manage framework-specific stuff. For example, if Jax users want to build TE/Jax, then PyTorch should not be required in their environment.

The user can set up an environment variable - `FRAMEWORK` to select framework, such as:
```sh
FRAMEWORK=jax pip install .
FRAMEWORK=pytorch pip install .
FRAMEWORK=all pip install .
```
The default value is `all`

The other Jax modules will be submitted after this PR is merged.

### CODE REVIEW

**Comment by jeng1220** [comment]
Addition, TE/Jax need pybind11 to build Python-CPP binding, so we need
```sh
apt install pybind11-dev
pip install pybind11
```
Both of above packages are needed

File: setup.py [hunk 1]
```diff
@@ -112,21 +113,42 @@ def __init__(self, name, cmake_path, sources, **kwargs):
...
+
+    def initialize_options(self):
+        self.pytorch_build_extensions.initialize_options()
+
+    def finalize_options(self):
+        self.pytorch_build_extensions.finalize_options()
+
+    def run(self):
+        from torch.utils.cpp_extension import CUDAExtension
+        other_ext = CUDAExtension(
```
Comment (timmoon10) [inline setup.py:151]: I'm not very familiar with `setuptools`, but I wonder if this is a cause of your build error. This `CUDAExtension` does not have a chance to initialize/finalize options before it is run. I wonder if you need to construct it before being passed into `setuptools.setup`: https://github.com/NVIDIA/TransformerEngine/blob/47f1a2910c628d852b9094ed2dc64f3ab63fe948/setup.py#L378

Comment (jeng1220) [reply setup.py]: This issue was solved.
Moving `CUDAExtension ` back to `ext_modules` can address the error.
I'm not very familiar with setuptools either. This is the best I can find for now.  :disappointed_relieved:

**Comment by jeng1220** [comment]
@timmoon10 ,
The CI triggers build failed:
https://github.com/NVIDIA/TransformerEngine/actions/runs/3901104182/jobs/6662679011#step:4:85

As I mentioned, TE/Jax needs pybind11
We have to install `pybind11-dev` in the CI container.
```sh
$ apt install pybind11-dev
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
Sorry for the delay. I think this is mostly fine. I could imagine some future build errors if, for example, the CMake flags for the PyTorch and JAX extensions are incompatible. That seems unlikely though, and we can refactor `setup.py` as any issues come up.

The biggest issues are merge conflicts. In particular, https://github.com/NVIDIA/TransformerEngine/pull/51 removes the `scale_inv` arg from many of the TE functions and makes it the responsibility of the frameworks. This will likely also affect https://github.com/NVIDIA/TransformerEngine/pull/54.

File: transformer_engine/jax/csrc/utils.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+namespace transformer_engine {
+namespace jax {
+
+template <class To, class From>
+typename std::enable_if<sizeof(To) == sizeof(From) &&
+                            std::is_trivially_copyable<From>::value &&
+                            std::is_trivially_copyable<To>::value,
+                        To>::type
+bit_cast(const From &src) noexcept {
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/csrc/utils.h]: Done

File: transformer_engine/jax/csrc/modules.cc [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+
+void TECastTranspose(cudaStream_t stream, void **buffers, const char *opaque,
+                     std::size_t opaque_len) {
+  const TEMatDescriptor &d =
+      *UnpackDescriptor<TEMatDescriptor>(opaque, opaque_len);
+
+  const void *input = reinterpret_cast<const void *>(buffers[0]);
+  const void *amax = reinterpret_cast<const void *>(buffers[1]);
+  float *scale = reinterpret_cast<float *>(buffers[2]);
+  float *scale_inv = reinterpret_cast<float *>(buffers[3]);
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/csrc/modules.cc:106]: It doesn't work. We still have to pass `scale_inv` to kernel although `scale_inv` won't be computed. The #51 only removed the computation but didn't the checking and API for `scale_inv`. For example: 

- https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/transformer_engine.cpp#L27
- https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/transformer_engine.cpp#L51

File: setup.py [hunk 4]
```diff
@@ -94,9 +95,10 @@ def make_abs_path(l):
 supported_frameworks = {
     "all": all_sources,
     "pytorch": pytorch_sources,
+    "jax": None, # JAX use transformer_engine/CMakeLists.txt
 }
 
-framework = "all"
+framework = os.environ.get("FRAMEWORK", "all")
```
Comment (ptrendx) [inline setup.py:101]: I would actually set that to pytorch for now in order to not break the existing workflows. We can switch that to all once the JAX support is fully merged.

**Comment by timmoon10** [comment]
Once we iron out the style and build issues, I think this is okay to merge.

I'm not thrilled with how we're refactoring layer norm and RMSNorm to use the same helper function. In principle they are independent operations that just happen to look similar. But in practice I don't expect them to undergo significant API changes in the future, and this is an internal implementation detail anyways.

File: setup.py (hunk 4 above)
Comment (jeng1220) [reply setup.py]: Done

File: transformer_engine/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+
+project(transformer_engine LANGUAGES CUDA CXX)
+
+list(APPEND CMAKE_CUDA_FLAGS "--threads 4")
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+  set(CMAKE_CUDA_FLAGS_DEBUG "${CMAKE_CUDA_FLAGS_DEBUG} -G")
+endif()
+
+find_package(CUDAToolkit REQUIRED cublas)
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/CMakeLists.txt]: Done

**Review by timmoon10 — APPROVED** [review]

**Review by timmoon10 — CHANGES_REQUESTED** [review]
I see the linter found some style issues.

**Comment by jeng1220** [comment]
> I see the linter found some style issues.

It should be done

File: transformer_engine/jax/cpp_extensions.py [hunk 6]
```diff
@@ -0,0 +1,1975 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+# pylint: disable=c-extension-no-member
+# pylint: disable=no-name-in-module
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/cpp_extensions.py:6]: I saw pylint reports that warning but didn't know how to resolve it even I google that warnings. Thus, I suppress them. Do you have expereinces on `c-extension-no-member` and `no-name-in-module`? Both them are regarding Python CPP binding.

**Comment by timmoon10** [comment]
The scope of this PR has grown quite a bit. I'm still working through everything, but I don't see any glaring issues yet.

/te-ci

File: transformer_engine/jax/cpp_extensions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/jax/cpp_extensions.py:6]: I'm just suspicious of disabling warnings if we don't understand what's causing them. It's just asking for problems to show up in the future.

Did you get any more info about where it's happening? It makes me think something isn't being imported correctly. I wonder if it's because pylint is running in an environment without JAX?

File: tests/jax/test_custom_call_shape.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                                              mat_b,
+                                              scale_inv_a,
+                                              scale_inv_b,
+                                              A_dtype=te_dtype,
+                                              B_dtype=te_dtype,
+                                              D_dtype=te_dtype,
+                                              transa=transa,
+                                              transb=transb,
+                                              use_split_accumulator=False)
+            assert ref_out == test_out
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply tests/jax/test_custom_call_shape.py:91]: Is `assert not need_assert` a pseudo code?

**Comment by jeng1220** [comment]
> The scope of this PR has grown quite a bit. I'm still working through everything, but I don't see any glaring issues yet.

Sorry for inconvenience.
We tried to refactor the code and separate them into smaller pieces.
But the schedule pression doesn't allow us to refactor further

File: transformer_engine/jax/cpp_extensions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/cpp_extensions.py:6]: I saw a workaround here: https://stackoverflow.com/questions/28437071/pylint-1-4-reports-e1101no-member-on-all-c-extensions.

In short reason:

> Pylint was decided that importing c extensions that are not part of the python stdlib is a security risk and could introduce malicious code.

The workaround:
`extension-pkg-whitelist=transformer_engine_jax` in `pylint.rc` to disable warning

[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/cpp_extensions.py:6]: Hmm, it is strange that why warnings were not triggered by TE/PyTorch

[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/cpp_extensions.py:6]: After trial and error for hours, I still don't know why warnings were not triggered by TE/PyTorch.
The pylint message also said:
```
transformer_engine/jax/cpp_extensions.py:24:21: I1101: Module 'transformer_engine_jax' has no 'registrations' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
```
Thus, I think adding extension-pkg-whitelist=transformer_engine_jax in pylint.rc is best approach for now.
Does it make sense?

**Comment by jeng1220** [comment]
@timmoon10 ,
I just rebased to `main`, and #62 breaks the building process for TE/JAX again.
Looking at it...

**Comment by jeng1220** [comment]
> 



>

**Comment by jeng1220** [comment]
> > @timmoon10 , I just rebased to `main`, and #62 breaks the building process for TE/JAX again. Looking at it...
> 
> The problem is here [setup.py#L317](https://github.com/NVIDIA/TransformerEngine/blob/main/setup.py#L317). It requires installing `flash-attn` and flash-attn requires installing PyTorch. However, usually, JAX users don't have PyTorch in their environment.

It should be fixed by 2321cc3

File: transformer_engine/jax/cpp_extensions.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/cpp_extensions.py:6]: @timmoon10 ,
This is why TE/PyTorch didn't trigger the warning: [qa/L0_lint/pylintrc#L2](https://github.com/NVIDIA/TransformerEngine/blob/main/qa/L0_lint/pylintrc#L2)

```python
extension-pkg-whitelist=torch,
                        transformer_engine_extensions
```
I will add `transformer_engine_jax` to `extension-pkg-whitelist` and push a commit later

[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/cpp_extensions.py]: This issue should be solved by [252b8bf](https://github.com/NVIDIA/TransformerEngine/pull/53/commits/252b8bfd953abc552f00d7d87a4e6a28abc68448)

[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/jax/cpp_extensions.py]: Nice debugging, this is a much cleaner solution.

File: setup.py [hunk 8]
```diff
@@ -94,9 +94,10 @@ def make_abs_path(l):
 supported_frameworks = {
     "all": all_sources,
     "pytorch": pytorch_sources,
+    "jax": None, # JAX use transformer_engine/CMakeLists.txt
 }
 
-framework = "all"
+framework = os.environ.get("FRAMEWORK", "pytorch")
```
Comment (ksivaman) [inline setup.py:100]:
```suggestion
framework = os.environ.get("NVTE_FRAMEWORK", "pytorch")
```

Comment (ksivaman) [reply setup.py:100]: Keeping it consistent with all other environment variables that TE uses

File: transformer_engine/__init__.py [hunk 9]
```diff
@@ -2,6 +2,19 @@
 #
 # See LICENSE for license information.
 
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+# pylint: disable=missing-class-docstring
```
Comment (ksivaman) [inline transformer_engine/__init__.py:7]: Why are these needed? Current version of __init__ passes without these disables and I don't see any change here that would change that

File: tests/jax/test_custom_call_compute.py [hunk 10]
```diff
@@ -0,0 +1,575 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (ksivaman) [inline tests/jax/test_custom_call_compute.py:1]: Since we have a clear separation of the 2 frameworks in TE, I think it also makes sense in this PR to add the PyTorch tests in a separate folder `tests/pytorch`. The ONNX stuff can be moved there as well. So the tests would have 2 directories (`jax` and `pytorch`) and no files.

File: tests/jax/test_custom_call_compute.py [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+#
+# See LICENSE for license information.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+# pylint: disable=missing-class-docstring
```
Comment (ksivaman) [inline tests/jax/test_custom_call_compute.py:7]: IMO the `tests` directory should either be immune to linting or we should actually comply to the rules. Having lint enabled here but bypassing certain flags does not seem right.

File: tests/jax/test_custom_call_shape.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+# pylint: disable=missing-class-docstring
```
Comment (ksivaman) [inline tests/jax/test_custom_call_shape.py:7]: Same as above.

File: tests/jax/test_helper.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+# pylint: disable=missing-class-docstring
```
Comment (ksivaman) [inline tests/jax/test_helper.py:7]: Same as above.

File: tests/jax/utils.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+# pylint: disable=missing-module-docstring
+# pylint: disable=missing-function-docstring
+# pylint: disable=missing-class-docstring
```
Comment (ksivaman) [inline tests/jax/utils.py:7]: Same as above.

File: tests/jax/test_custom_call_compute.py (hunk 10 above)
Comment (timmoon10) [reply tests/jax/test_custom_call_compute.py:1]: We'll also need to adjust paths at https://github.com/jeng1220/TransformerEngine/blob/rjeng/init_jax/qa/L0_unittest/test.sh.

**Comment by timmoon10** [comment]
The test failures are unrelated to this PR (see https://github.com/NVIDIA/TransformerEngine/pull/67#issuecomment-1430626755), so I think this is ready after addressing @ksivaman's concerns. I don't have full confidence that I got everything and in the future I would try to keep each PR with a clearly defined scope. It's much faster and safer to review multiple simple PRs than a single monstrous one.

File: transformer_engine/__init__.py (hunk 9 above)
Comment (jeng1220) [reply transformer_engine/__init__.py]: It is caused by misaligned pylint setting between `main` and internal devel branch.
Now, it should be fixed by 5615c9b

File: tests/jax/test_custom_call_compute.py (hunk 10 above)
Comment (jeng1220) [reply tests/jax/test_custom_call_compute.py:1]: I can help to submit another PR to move files regarding PyTorch into `tests/pytorch`. Hope to focus on TE/JAX in this PR

File: transformer_engine/jax/sharding.py [hunk 15]
```diff
@@ -0,0 +1,902 @@
...
+    `TP_ROW` means sharding along row-split tensor parallelism.
+    `DP_TP_COL` means sharding along data and column-split tensor parallelism.
+    `DP_TP_ROW` means sharding along data and row-split tensor parallelism.
+    """
+    SINGLE = (MajorShardingType.SINGLE, "single")
+    DP = (MajorShardingType.DP, "dp")
+    TP_COL = (MajorShardingType.TP, "tp_col")
+    TP_ROW = (MajorShardingType.TP, "tp_row")
+    DP_TP_COL = (MajorShardingType.DPTP, "dp_tp_col")
+    DP_TP_ROW = (MajorShardingType.DPTP, "dp_tp_row")
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply transformer_engine/jax/sharding.py:102]: This is an action item in our schedule. But to refactor this part, probably will take more than 2 weeks. That will be too late to release TE/JAX.

Those `sharding` stuff is low-level implementation in TE/JAX. The TE/JAX users don't and shouldn't cal them directly. In the next PR, we will provide high-level module to hide these stuff.

I fully agree this part should be better. We will continue refactoring after first TE/JAX release is done.

File: tests/jax/test_custom_call_compute.py (hunk 10 above)
Comment (jeng1220) [reply tests/jax/test_custom_call_compute.py:1]: @ksivaman and @timmoon10 ,
It should be solved by https://github.com/NVIDIA/TransformerEngine/pull/78

**Comment by jeng1220** [comment]
> The test failures are unrelated to this PR (see [#67 (comment)](https://github.com/NVIDIA/TransformerEngine/pull/67#issuecomment-1430626755)), so I think this is ready after addressing @ksivaman's concerns. I don't have full confidence that I got everything and in the future I would try to keep each PR with a clearly defined scope. It's much faster and safer to review multiple simple PRs than a single massive one.

@timmoon10 ,
I understand it is difficult to review so much files at once.
The next PR will include less files. Hope it can be easier for code review.
I appreciate your help very much.

**Review by timmoon10 — APPROVED** [review]

**Comment by jeng1220** [comment]
@ksivaman and @timmoon10 ,
I have merged the `main`

**Comment by timmoon10** [comment]
@ksivaman and I think this is ready to merge, but we'd like to discuss with @ptrendx tomorrow. Barring any issues, we'll merge it by California afternoon.
